File: LeetCode/p121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        invest = prices[0]
        for i in range(1, len(prices)):
            if prices[i] < invest:
                invest = prices[i]
            else:
                profit = max(prices[i] - invest, profit)

        return profit



if __name__ == '__main__':
    sol = Solution()
    print(sol.maxProfit([2,4,1]))

File: LeetCode-Java/p121.java
import java.util.List;

class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }
        
        int profit = 0;
        int invest = prices[0];
        
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < invest) {
                invest = prices[i];
            } else {
                profit = Math.max(prices[i] - invest, profit);
            }
        }
        
        return profit;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] prices = {2, 4, 1};
        System.out.println(sol.maxProfit(prices));
    }
}